lib/model/model.go

```diff
@@ -2606,3 +2627,37 @@ func (s folderDeviceSet) sortedDevices(folder string) []protocol.DeviceID {
 	sort.Sort(protocol.DeviceIDs(devs))
 	return devs
 }
+
+// rootedJoinedPath takes a root and a supposedly relative path inside that
+// root and returns the joined path. An error is returned if the joined path
+// is not in fact inside the root.
+func rootedJoinedPath(root, rel string) (string, error) {
+	// The expected prefix for the resulting path is the root, with a path
+	// separator at the end.
+	expectedPrefix := filepath.FromSlash(root)
+	if !strings.HasSuffix(expectedPrefix, string(os.PathSeparator)) {
+		expectedPrefix += string(os.PathSeparator)
+	}
+
+	// The relative path should be clean from internal dotdots and similar
+	// funkyness.
+	rel = filepath.FromSlash(rel)
+	if filepath.Clean(rel) != rel {
+		return "", errInvalidFilename
+	}
+
+	// It is not acceptable to attempt to traverse upwards.
+	if strings.HasPrefix(rel, ".."+string(os.PathSeparator)) {
```

Comment:
It's tested for though, gets caught by the prefix check below. This check is needed anyway to catch `join("foo/bar", "../bar/baz")` which becomes "foo/bar/baz" (but should still not be allowed).